### DISCUSSION TITLE
Intellisense features for Java packages using the Jython Interpreter

### DISCUSSION THREAD

**Discussion by Khalid-Tarek** [microsoft/pylance-release#6422] [Q&A]
I have been trying to setup an environment where I can create scripts for IBM Maximo using Jython.  I have downloaded the Jython interpreter, added the maximo library to the Jython library, and finally selected Jython as my Python interpreter.

The following behavior arises: -
- Libraries/Packages are recognized, I am assuming that this is already a feature in this extension:
![jython-library-autocompletion](https://github.com/user-attachments/assets/e11d299c-9137-4086-a6b0-91f490c81d59)
- On the other hand, there are no suggestions for any classes inside the packages. This happens with both .class and the decompiled .java files:
![jython-autocompletion-missing](https://github.com/user-attachments/assets/3c4cba8f-1275-40f8-b98d-dca483d5ce8c)

My end goal is to enable the following features: -
- code completion (listing an object's functions, attributes, etc for example) and 
- show hover (view help/javadocs for the java elements that have javadocs for example)

So far, after searching for an existing extension that adds Jython support to VSCode, I have found none. Therefore, the only solution I can see right now is to create an extension that achieves that. I am not fluent in VSCode API yet, and before I commit to learning it, I wanted to inquire about the validity of my thought process:
1. Learn how vscode-java implements autocompletion, hover and maybe other programmatic language features
2. Similarly learn how vscode-python implements the same features
3. Integrate the way vscode-java implements these features into vscode-python

I am assuming it would be much easier if the library is in the decompiled form (.java files) but I am also going to look into using the singular .jar library as well. I would also appreciate any comments on that.

**Comment by karthiknadig** [comment]
Moving this to Pylance.

There are two approaches to python completion, Jedi (object inspection), and Pylance (Type Analysis). Jedi Language Server: https://github.com/pappasam/jedi-language-server uses `jedi` library to do this.

**Reply by Khalid-Tarek** [reply]
I just read into pylance and it states that pylance is not open source, but states that there's another project called pyright that is open source that I can work with. As for my approach here, if I am looking into adding jython support to vscode, should I think of it as a seperate extension or extend another extension (like pylance / pyright for example)?

As for Jedi Language Server, it is unrelated to pylance is that correct? as in its a separate extension.

Apologies if my questions seem very elementary, I am still in the process of learning how extensions / language servers work

**Reply by karthiknadig** [reply]
You would create a dedicated language server that specifically supports Jython. But you can use many parts of pyright to build it.

Jedi is a different project, it also provides intellisence features. The approach it takes is very different. It loads the file in the Python runtime and examines them. Pyright, uses the parse tree to examine the Python files. The two approaches have different advantages and disadvantages.

Jedi approach is a simpler approach. But one thing to keep in mind, I the the latest stable for jython is still Python 2.7 syntax. And both pyright and Jedi don’t support that any more.


For your project you can build on top of either pyright or Jedi.

**Reply by Khalid-Tarek** [reply]
Python 2.7 not being supported is honestly a bit disheartening and could be a hurdle to my plan, but like I said I am not really looking to compile the code whatsoever, only to facilitate autocomplete mainly and other intellisense features as a secondary goal. Since you say that Jedi loads the file in python runtime, I am guessing I will have a lot of issues if I go down that route, do you agree? 

As for creating my own dedicated language server; as an additional goal, I would like to contribute whatever I reach to the community (i.e. the 2 or 3 other people that use jython 😅). Whether that's by creating a new vscode extension from the ground up (what I prefer), or through a pull request to pyright / jedi, it's not really a concern to me. My end goal is the functionality itself.

What I am worried about is how the language server detects elements inside the file, as in it's classes, methods, properties, etc... Is that even possible to parse a Java file from pyright? As in, the way a Java language server would? or would I have to implement my own Java file parser in pyright somehow?

**Reply by rchiodo** [reply]
You'd have to implement your own java file parser. If jython is about parsing java, Pyright may not be the best example. It's strictly around parsing python files and computing their types.

**Reply by Khalid-Tarek** [reply]
According to their [website](https://www.jython.org/):
> The Jython project provides implementations of [Python](https://www.python.org/) in [Java](https://go.java/index.html), providing to Python the benefits of running on the JVM and access to classes written in Java.

Since it is an open source project, could it be worthwhile to investigate how importing of Java files work in python code and how that could translate to parsing these Java files in a dedicated language server, or would that be a waste of time? I am guess that could be a waste of time perhaps because language servers might have nothing to do with actual compilation and execution  of the code, if I am understanding correctly.

In that case, I am thinking of doing the following:
- Finding an open source Java language server, and understand how Java files are parsed there
- Understanding how python files are parsed in pyright
- Implementing the Java file parsing mentioned in the first point, in addition to a fork of pyright / a new extension (apologies if I am using the terminology wrong, this is my first discussion on github)
